### ISSUE TITLE
Mean-Average-Precision gives unexpected result.

### ISSUE THREAD

**Issue by kfarivar** [Lightning-AI/torchmetrics#1966]
## 🐛 Bug

Changing the order of bounding boxes in Mean-Average-Precision changes the result.

### To Reproduce

Steps to reproduce the behavior...

```python
from torch import tensor
from torchmetrics.detection import MeanAveragePrecision
preds = [
   dict(
     boxes=tensor([  [214.0, 41.0, 562.0, 285.0], [600.0, 300.0, 900., 600.],  ]),
     scores=tensor([1., 1.]),
     labels=tensor([1, 1]),
   )
 ]
target = [
   dict(
     boxes=tensor([[214.0, 41.0, 562.0, 285.0]]),
     labels=tensor([1]),
   )
 ]
metric = MeanAveragePrecision() 
metric.update(preds, target)
from pprint import pprint
pprint(metric.compute())
```

switching the two boxes in preds changes the result from 1 to .5

### Expected behavior

the results should be independent of box order. 

### Environment

- TorchMetrics version: 1.0.1  using pip
- Python: 3.10.10 & PyTorch: 2.0.1  
- running on linux: "Debian GNU/Linux 11 (bullseye)"

**Comment by SkafteNicki** [comment]
Duplicate of: https://github.com/Lightning-AI/torchmetrics/issues/1774
TLDR: the result are not unexpected and that is the exact same behaviour the official implementation would have.

Long answer:
The core problem is that to calculate the MAP, we first need to calculate the precision-recall-curve. This is done by first sorting our predictions based on their confidence. This is a problem when predictions have the exact same confidence (as in your example) then this sorting is not unique and therefore depends on the order of the input.

If we calculate the precision recall curve for the two cases:

*correspond to the wrong prediction being first*
Conf | Matches | Cumulative TP | Cumulative FP | Precision | Recall
-- | -- | -- | -- | -- | --
1.0 | TP | 0 | 1 | 0/(0+1) = 0 | 0/1=0
1.0 | FP | 1 | 1 | 1/(1+1) = 1/2 | 1/1=1

*correspond to the right prediction being first*
Conf | Matches | Cumulative TP | Cumulative FP | Precision | Recall
-- | -- | -- | -- | -- | --
1.0 | TP | 1 | 0 | 1/(1+0) = 1 | 1/1=1
1.0 | FP | 1 | 1 | 1/(1+1) = 1/2 | 1/1=1

Afterwards we would interpolate the two curves and calculate the area under them which would give us the MAP value. I wont go into details on how to this, but it should not be surprising that if the two curves are different then the area under them will also be different. For the first case you end up with a 0.5 value and the second case a value of 1, as you have already seen.

**Comment by kfarivar** [comment]
Thank you for the explanation. I still have problem understanding the results. 
In the case for _right prediction being first_ shouldn't the extra false positive prediction reduce the AP ?
to be clear we can give the wrong prediction score of 0.9.

**Comment by SkafteNicki** [comment]
@kfarivar no problem. I try with more details:

Your original question was regarding why the input order mattered for the score the metric produced. Hopefully, my last answer makes it clear that it has to do with the the two predictions having equal confidence leading to the input order having a influence when we calculate the precision recall curve.
Thus if we change the wrong prediction to have a 0.9 confidence we will correctly see that the input order does not matter, the result will be 1.0 in both cases e.g

```python 
from torch import tensor
from torchmetrics.detection import MeanAveragePrecision
from pprint import pprint
preds = [
   dict(     boxes=tensor([  [214.0, 41.0, 562.0, 285.0], [600.0, 300.0, 900., 600.],  ]),
     scores=tensor([1., 0.9]),
     labels=tensor([1, 1]),
   )
 ]
# change order for both boxes and confidences
preds2 = [
   dict(     boxes=tensor([  [600.0, 300.0, 900., 600.], [214.0, 41.0, 562.0, 285.0] ]),
     scores=tensor([0.9, 1.0]),
     labels=tensor([1, 1]),
   )
 ]

target = [
   dict(
     boxes=tensor([[214.0, 41.0, 562.0, 285.0]]),
     labels=tensor([1]),
   )
 ]
metric = MeanAveragePrecision()
pprint(metric(preds, target))  # 1.0 in map
pprint(metric(preds2, target))  # also 1.0 in map
```
Regarding your new question:
> In the case for right prediction being first shouldn't the extra false positive prediction reduce the AP ? o/w the model with lots of false positives wouldn't be any different from one without ?

I understand if this is counter intuitive, but the core thing for this particular simple case is that if we match the target box on the first prediction, then it will not matter how many we get wrong afterwards. 
If we take this prediction for example:

```python
preds3 = [
   dict(     boxes=tensor([
                [600.0, 300.0, 900., 600.],
                [600.0, 300.0, 900., 600.],
                [600.0, 300.0, 900., 600.],
                [600.0, 300.0, 900., 600.],
                [600.0, 300.0, 900., 600.],
                [214.0, 41.0, 562.0, 285.0] ]),
     scores=tensor([0.9, 0.9, 0.9, 0.9, 0.9, 1.0]),
     labels=tensor([1, 1, 1, 1, 1, 1]),
   )
 ]
```
And repeat the precision recall calculation: 

Conf | Matches | Cumulative TP | Cumulative FP | Precision | Recall
-- | -- | -- | -- | -- | --
1.0 | TP | 1 | 0 | 1/(1+0) = 1 | 1/1=1
0.9 | FP | 1 | 1 | 1/(1+1) = 1/2 | 1/1=1
0.9 | FP | 1 | 2 | 1/(1+2) = 1/3 | 1/1=1
0.9 | FP | 1 | 3 | 1/(1+3) = 1/4 | 1/1=1
0.9 | FP | 1 | 4 | 1/(1+4) = 1/5 | 1/1=1
0.9 | FP | 1 | 5 | 1/(1+6) = 1/6 | 1/1=1

Importantly the recall value is 1.0 regardless of how many false positives we observe after the first true positive. 
Here is a quick paint drawing, where the red points correspond to the numbers in the table and the green line is the curve we would interpolate under:
![Untitled](https://github.com/Lightning-AI/torchmetrics/assets/24896311/3134c4a9-bf26-45cd-bdeb-c9531258ae63)

As you can see, it does not matter how many false positives we add because they all are on the line we already are using for interpolation.

Hope this clears it up a bit more :]
